src/mystery_gift.c

```diff
@@ -0,0 +1,293 @@
+        if (!MysteryGiftTagIsValid(mg->gifts[i].tag)) {
+            return TRUE;
+        }
+    }
+
+    return FALSE;
+}
+
+BOOL SaveMysteryGift_CardFindAvailable(const MysteryGiftSave* mg) {
+    int i;
+
+    for (i = 0; i < NUM_SAVED_WONDER_CARDS; ++i) {
+        if (!MysteryGiftTagIsValid(mg->cards[i].tag)) {
+            return TRUE;
+        }
+    }
+
+    return FALSE;
+}
+
+BOOL SaveMysteryGift_TagAtIndexIsValid(const MysteryGiftSave* mg, int index) {
+    GF_ASSERT(index < NUM_SAVED_MYSTERY_GIFTS);
+    
+    if (MysteryGiftTagIsValid(mg->gifts[index].tag)) {
+        return TRUE;
+    } else {
+        return FALSE;
+    }
+}
+
+BOOL SaveMysteryGift_CardTagIsValid(const MysteryGiftSave* mg, int index) {
+    GF_ASSERT(index < NUM_SAVED_WONDER_CARDS);
+    
+    if (MysteryGiftTagIsValid(mg->cards[index].tag)) {
+        return TRUE;
+    } else {
+        return FALSE;
+    }
+}
+
+BOOL SaveMysteryGift_SpecialCardTagIsValid(const MysteryGiftSave* mg) {
+    if (MysteryGiftTagIsValid(mg->specialWonderCard.tag)) {
+        return TRUE;
+    } else {
+        return FALSE;
+    }
+}
+
+BOOL SaveMysteryGift_HasAnyCard(const MysteryGiftSave* mg) {
+    int i;
+
+    for (i = 0; i < NUM_SAVED_WONDER_CARDS; ++i) {
+        if (SaveMysteryGift_CardTagIsValid(mg, i) == TRUE) {
+            return TRUE;
+        }
+    }
+
+    return FALSE;
+}
+
+BOOL SaveMysteryGift_HasAnyGift(const MysteryGiftSave* mg, int a1) {
+    int i;
+
+    for (i = 0; i < NUM_SAVED_MYSTERY_GIFTS; ++i) {
+        if (MysteryGiftTagIsValid(mg->gifts[i].tag) && a1 == mg->gifts[i].flag) {
+            return TRUE;
+        }
+    }
+
+    return FALSE;
+}
+
+void SaveMysteryGift_SetReceivedByCardId(MysteryGiftSave* mg, int a1) {
+    int i;
+
+    for (i = 0; i < NUM_SAVED_MYSTERY_GIFTS; ++i) {
+        if (MysteryGiftTagIsValid(mg->gifts[i].tag) && a1 == mg->gifts[i].flag) {
+            SaveMysteryGift_SetReceivedByIdx(mg, i);
+            break;
+        }
+    }
+}
+
+BOOL SaveMysteryGift_ReceivedFlagTest(const MysteryGiftSave* mg, int index) {
+    GF_ASSERT(index < NUM_MYSTERY_GIFT_RECV_FLAGS);
+    return mg->receivedFlags[index / 8] & (1 << (index & 7)) ? TRUE : FALSE;
+}
+
+void SaveMysteryGift_ReceivedFlagSet(MysteryGiftSave* mg, int index) {
+    GF_ASSERT(index < NUM_MYSTERY_GIFT_RECV_FLAGS);
+    mg->receivedFlags[index / 8] |= (1 << (index & 7));
+    SaveSubstruct_UpdateCRC(SAVE_MYSTERY_GIFT);
+}
+
+void SaveMysteryGift_ReceivedFlagClear(MysteryGiftSave* mg, int index) {
+    u8 mask = ~(1 << (index & 7));
+    GF_ASSERT(index < NUM_MYSTERY_GIFT_RECV_FLAGS);
+    mg->receivedFlags[index / 8] &= mask;
+    SaveSubstruct_UpdateCRC(SAVE_MYSTERY_GIFT);
+}
+
+BOOL SaveMysteryGift_TestFlagx7FF(const MysteryGiftSave* mg) {
+    return SaveMysteryGift_ReceivedFlagTest(mg, 0x7FF);
+}
+
+void SaveMysteryGift_SetFlagx7FF(MysteryGiftSave* mg) {
+    SaveMysteryGift_ReceivedFlagSet(mg, 0x7FF);
+}
+
+static MysteryGiftSave* sMysteryGiftData;
```

Comment:
personally I'd prefer if all vars were at the top of the file, but it's not 100% necessary. Also might be better to make it sMysteryGiftSaveData